src/cryptonote_core/service_node_quorum_cop.cpp

```diff
@@ -176,19 +176,61 @@ namespace service_nodes
               //
               // NOTE: I am in the quorum
               //
-              for (size_t node_index = 0; node_index < quorum->workers.size(); ++node_index)
+              auto worker_states = m_core.get_service_node_list_state(quorum->workers);
+              auto worker_it = worker_states.begin();
+              CRITICAL_REGION_LOCAL(m_lock);
+              for (size_t node_index = 0; node_index < quorum->workers.size(); ++worker_it, ++node_index)
               {
-                const crypto::public_key &node_key = quorum->workers[node_index];
-
-                CRITICAL_REGION_LOCAL(m_lock);
-                bool vote_off_node = (m_uptime_proof_seen.find(node_key) == m_uptime_proof_seen.end());
-                if (!vote_off_node) continue;
-
-                quorum_vote_t vote = service_nodes::make_deregister_vote(
-                    m_uptime_proof_height, static_cast<uint16_t>(index_in_group), node_index, my_pubkey, my_seckey);
+                // If the SN no longer exists then it'll be omitted from the worker_states vector,
+                // so if the elements don't line up skip ahead.
+                while (worker_it->pubkey != quorum->workers[node_index] && node_index < quorum->workers.size())
```

Comment:
Each piecemeal lookup would involve taking out a service node list lock, which would be rather heavy, but worse, you'd have a potential deadlock situation because there is *also* a lock on the quorum cop already held over this loop.  It would probably end up more complicated to go there.